File: ros2/imx708_camera_cpp/src/imx708_publisher.cpp
// IMX708 publisher for ROS 2 (C++).
// Captures via GStreamer/Argus, publishes sensor_msgs/Image on /image_raw.

#include <chrono>
#include <memory>
#include <sstream>
#include <string>

#include <opencv2/opencv.hpp>
#include <cv_bridge/cv_bridge.h>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>

using namespace std::chrono_literals;

namespace imx708_camera_cpp
{

static std::string build_pipeline(int width, int height, int fps, int sensor_id)
{
  std::ostringstream oss;
  oss << "nvarguscamerasrc sensor-id=" << sensor_id << " "
      << "wbmode=5 awblock=true "
      << "exposuretimerange=\"8000000 16000000\" "
      << "gainrange=\"1 4\" ispdigitalgainrange=\"1 1\" "
      << "! video/x-raw(memory:NVMM),width=" << width
      << ",height=" << height
      << ",framerate=" << fps << "/1,format=NV12 "
      << "! nvvidconv ! video/x-raw,format=BGRx "
      << "! videoconvert ! video/x-raw,format=BGR "
      << "! appsink drop=true max-buffers=2 sync=false";
  return oss.str();
}

class IMX708Publisher : public rclcpp::Node
{
public:
  IMX708Publisher()
  : Node("imx708_publisher")
  {
    width_     = this->declare_parameter<int>("width", 1536);
    height_    = this->declare_parameter<int>("height", 864);
    fps_       = this->declare_parameter<int>("fps", 30);
    sensor_id_ = this->declare_parameter<int>("sensor_id", 0);
    frame_id_  = this->declare_parameter<std::string>("frame_id", "imx708_optical_frame");

    const auto pipeline = build_pipeline(width_, height_, fps_, sensor_id_);
    RCLCPP_INFO(this->get_logger(), "Opening pipeline: %s", pipeline.c_str());

    cap_.open(pipeline, cv::CAP_GSTREAMER);
    if (!cap_.isOpened()) {
      RCLCPP_FATAL(this->get_logger(), "Failed to open GStreamer pipeline");
      throw std::runtime_error("VideoCapture::open failed");
    }

    pub_ = this->create_publisher<sensor_msgs::msg::Image>("image_raw", 10);

    const auto period = std::chrono::milliseconds(1000 / fps_);
    timer_ = this->create_wall_timer(period, std::bind(&IMX708Publisher::tick, this));

    RCLCPP_INFO(this->get_logger(),
      "Publishing %dx%d @ %d fps on /image_raw", width_, height_, fps_);
  }

  ~IMX708Publisher() override
  {
    if (cap_.isOpened()) {
      cap_.release();
    }
  }

private:
  void tick()
  {
    cv::Mat frame;
    if (!cap_.read(frame) || frame.empty()) {
      RCLCPP_WARN(this->get_logger(), "Frame grab failed");
      return;
    }

    std_msgs::msg::Header hdr;
    hdr.stamp = this->now();
    hdr.frame_id = frame_id_;

    auto msg = cv_bridge::CvImage(hdr, "bgr8", frame).toImageMsg();
    pub_->publish(*msg);

    if (++frame_count_ % static_cast<uint64_t>(fps_) == 0) {
      RCLCPP_DEBUG(this->get_logger(), "Published %lu frames",
        static_cast<unsigned long>(frame_count_));
    }
  }

  int width_, height_, fps_, sensor_id_;
  std::string frame_id_;
  cv::VideoCapture cap_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  uint64_t frame_count_ {0};
};

}  // namespace imx708_camera_cpp

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<imx708_camera_cpp::IMX708Publisher>());
  rclcpp::shutdown();
  return 0;
}

File: ros2/imx708_camera_py/imx708_camera_py/imx708_publisher.py
#!/usr/bin/env python3
"""IMX708 publisher for ROS 2 (Python).

Captures frames from the Arducam IMX708 via GStreamer/Argus and publishes
them as sensor_msgs/Image on /image_raw. Color correction belongs to
downstream nodes; the Argus ISP tuning for IMX708 is imperfect.
"""
import cv2
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge


def build_pipeline(width, height, fps, sensor_id=0):
    return (
        f"nvarguscamerasrc sensor-id={sensor_id} "
        f"wbmode=5 awblock=true "
        f'exposuretimerange="8000000 16000000" '
        f'gainrange="1 4" ispdigitalgainrange="1 1" '
        f"! video/x-raw(memory:NVMM),width={width},height={height},"
        f"framerate={fps}/1,format=NV12 "
        f"! nvvidconv ! video/x-raw,format=BGRx "
        f"! videoconvert ! video/x-raw,format=BGR "
        f"! appsink drop=true max-buffers=2 sync=false"
    )


class IMX708Publisher(Node):
    def __init__(self):
        super().__init__("imx708_publisher")
        self.declare_parameter("width", 1536)
        self.declare_parameter("height", 864)
        self.declare_parameter("fps", 30)
        self.declare_parameter("sensor_id", 0)
        self.declare_parameter("frame_id", "imx708_optical_frame")

        w = self.get_parameter("width").value
        h = self.get_parameter("height").value
        fps = self.get_parameter("fps").value
        sid = self.get_parameter("sensor_id").value
        self.frame_id = self.get_parameter("frame_id").value

        pipeline = build_pipeline(w, h, fps, sid)
        self.get_logger().info(f"Opening pipeline: {pipeline}")

        self.cap = cv2.VideoCapture(pipeline, cv2.CAP_GSTREAMER)
        if not self.cap.isOpened():
            raise RuntimeError("Failed to open IMX708 GStreamer pipeline")

        self.bridge = CvBridge()
        self.pub = self.create_publisher(Image, "image_raw", 10)
        self.timer = self.create_timer(1.0 / float(fps), self.tick)
        self.get_logger().info(f"Publishing {w}x{h} @ {fps} fps on /image_raw")

    def tick(self):
        ok, frame = self.cap.read()
        if not ok or frame is None:
            self.get_logger().warn("Frame grab failed")
            return
        msg = self.bridge.cv2_to_imgmsg(frame, encoding="bgr8")
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = self.frame_id
        self.pub.publish(msg)

    def destroy_node(self):
        if self.cap.isOpened():
            self.cap.release()
        return super().destroy_node()


def main():
    rclpy.init()
    node = IMX708Publisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
